packages/next/src/server/app-render/app-render.tsx

```diff
@@ -2068,3 +2594,72 @@ async function prerenderToStream(
     }
   }
 }
+
+const loadingChunks: Set<Promise<unknown>> = new Set()
+const chunkListeners: Array<(x?: unknown) => void> = []
```

Comment:
why do these need to be global? are we never going to e.g. do this for two renders at once? (in which case AFAICT we'd mix up their chunkListeners)?

Reply:
We render multiple pages in parallel but each chunk is only loaded once in turbopack b/c React caches the module resolving promise internally. So if chunk A begins loading in page A and page B also loads it there won't be another call to the loader observable from within this file. So the only way to guarantee that we always load everything needed ahead of an SSR render is to wait globally for all pending chunks loading to finish before moving on to rendering the page. It's not ideal but it's not worth a "better" hack when we probably just need support for this in Turbopack.

FWIW this isn't a practical issue for webpack b/c on the server all the chunks are already loaded eagerly

Reply:
and addressing this directly
>in which case AFAICT we'd mix up their chunkListeners
Yeah we do but that's sort of the point. every render's chunk listeners wait for every other render's chunks to load